docs/kb/semgrep-ci/azure-using-templates-with-semgrep.md

```diff
@@ -0,0 +1,75 @@
+---
+tags:
+  - Azure
+  - Templates
+description: Using templates in Azure pipelines with Semgrep commands.
+---
+# Using templates in Azure pipelines with Semgrep
+
+## Motivation
+
+Complex code can take more work to maintain and modify. So, encapsulating some functionalities in functions, methods, or external files is a good programming practice.
+That is the primary goal of Azure templates: extract some functionalities (snippets) defined in a big yaml file to an external file (template). With that, we are improving pipelines' readability and maintainability.
+
+## Defining Semgrep snippets in a template file
+
+Adding semgrep commands in a yaml template is a simple task. In the example, we will create two templates, one for Semgrep full scan and one for Semgrep pull request scans.
+
+Example for Semgrep full scan in a template yaml file:
+``````
+steps:
+- script: |
+    echo "Semgrep full scan"
+    python -m pip install --upgrade pip
+    pip install semgrep
+    semgrep ci
+``````
+
+Example for Semgrep pull request scan in a template yaml file:
+``````
+steps:
+- checkout: self
+  clean: true
+  fetchDepth: 10000
+- script: |
+    echo "Pull Request Scan from branch: $(Build.SourceBranchName)"
+    git fetch origin master:origin/master
+    python -m pip install --upgrade pip
+    pip install semgrep
+    semgrep ci
+  env:
+    SEMGREP_PR_ID: $(System.PullRequest.PullRequestNumber)
+    SEMGREP_BASELINE_REF: 'origin/master'
+``````
+
+> Note that we split semgrep scans (full and pull request) as there are some essential differences in setting environment variables such as SEMGREP_PR_ID and SEMGREP_BASELINE_REF in the pull request template.
+
+## Including templates in an Azure Pipeline
+
+Now we can add these templates in the repo, for example, in a folder called templates.
```

Comment:
Do we have to do this per repo that we want to onboard? Just checking because one of the benefits is code reuse, so I thought we could reference it "globally" across all repos. Thank you.

Reply:
Maybe it is better to say: "We can add these templates in the repo, for example, in a templates folder or even better, add them to a (centralised repo and reference it)[https://learn.microsoft.com/en-us/azure/devops/pipelines/process/templates?view=azure-devops#use-other-repositories]"

Reply:
It was reviewed by @sinat101 too.